File: swift/Sources/RiffOpenAIRealtime/WebSocketTransport.swift
import Foundation
import RiffCore

/// Anything that can carry realtime events. The WebSocket implementation below is the one that
/// needs no media stack; a WebRTC transport plugs in here without the provider changing.
public protocol RiffTransport: Sendable {
    func send(_ message: JSONValue) throws
    func sendAudio(_ pcm: Data)
    func close(reason: String?) async
}

/// WebSocket transport over `URLSessionWebSocketTask`, where audio is base64 PCM inside JSON events.
///
/// WebRTC handles packet loss and jitter better on a phone, but it needs a media stack. This
/// transport needs nothing beyond Foundation, which makes it the one that works everywhere.
public final class WebSocketTransport: RiffTransport, @unchecked Sendable {
    private let task: URLSessionWebSocketTask
    private let onMessage: @Sendable (JSONValue) -> Void
    private let onClose: @Sendable (String?) -> Void
    private let onError: @Sendable (Error) -> Void
    private let lock = NSLock()
    private var closed = false

    public static let defaultURL = URL(string: "wss://api.openai.com/v1/realtime")!

    public init(
        url: URL = WebSocketTransport.defaultURL,
        model: String,
        token: String,
        organization: String? = nil,
        project: String? = nil,
        urlSession: URLSession = .shared,
        onMessage: @escaping @Sendable (JSONValue) -> Void,
        onClose: @escaping @Sendable (String?) -> Void,
        onError: @escaping @Sendable (Error) -> Void
    ) {
        self.task = urlSession.webSocketTask(with: WebSocketTransport.request(url: url, model: model, protocols: {
            var protocols = ["realtime", "openai-insecure-api-key.\(token)"]
            if let organization { protocols.append("openai-organization.\(organization)") }
            if let project { protocols.append("openai-project.\(project)") }
            return protocols
        }()))

        self.onMessage = onMessage
        self.onClose = onClose
        self.onError = onError
    }

    /// Adds the model without disturbing parameters the caller's endpoint already carries.
    ///
    /// Azure and gateway endpoints require parameters such as `api-version` and `deployment`.
    /// Replacing the query wholesale drops them, which fails in a way that looks like an auth error.
    static func request(url: URL, model: String, protocols: [String]) -> URLRequest {
        var components = URLComponents(url: url, resolvingAgainstBaseURL: false)
        var items = components?.queryItems ?? []
        items.removeAll { $0.name == "model" }
        items.append(URLQueryItem(name: "model", value: model))
        components?.queryItems = items

        var request = URLRequest(url: components?.url ?? url)
        request.setValue(protocols.joined(separator: ", "), forHTTPHeaderField: "Sec-WebSocket-Protocol")
        return request
    }

    public func resume() {
        task.resume()
        receive()
    }

    private func receive() {
        task.receive { [weak self] result in
            guard let self else { return }
            switch result {
            case .success(let message):
                let text: String? = switch message {
                case .string(let value): value
                case .data(let data): String(data: data, encoding: .utf8)
                @unknown default: nil
                }
                if let text, let data = text.data(using: .utf8),
                   let value = try? JSONDecoder().decode(JSONValue.self, from: data) {
                    self.onMessage(value)
                }
                self.receive()

            case .failure(let error):
                self.lock.lock()
                let wasClosed = self.closed
                self.lock.unlock()
                if !wasClosed {
                    self.onError(error)
                    self.onClose(error.localizedDescription)
                }
            }
        }
    }

    public func send(_ message: JSONValue) throws {
        task.send(.string(message.serialized())) { [weak self] error in
            if let error { self?.onError(error) }
        }
    }

    public func sendAudio(_ pcm: Data) {
        try? send(.object([
            "type": .string(OpenAIClientEvent.appendAudio),
            "audio": .string(pcm.base64EncodedString()),
        ]))
    }

    public func close(reason: String?) async {
        markClosed()
        task.cancel(with: .goingAway, reason: reason?.data(using: .utf8))
        onClose(reason)
    }

    // NSLock cannot be taken from an async context, so the critical section stays synchronous.
    private func markClosed() {
        lock.lock()
        closed = true
        lock.unlock()
    }
}

File: packages/riff-openai-realtime/src/websocket-transport.ts
import type { CredentialProvider, RealtimeTransport } from "./transport.ts";
import { encodeBase64, endpointWithModel } from "./transport.ts";
import { CLIENT_EVENTS } from "./events.ts";

export interface WebSocketLike {
  readyState: number;
  send(data: string): void;
  close(code?: number, reason?: string): void;
  addEventListener(type: string, listener: (event: any) => void): void;
}

export interface WebSocketTransportOptions {
  url?: string;
  model: string;
  credentials: CredentialProvider;
  organization?: string;
  project?: string;
  /** Supply your own socket to add headers, a proxy, or a custom TLS configuration. */
  createWebSocket?: (url: string, protocols: string[]) => WebSocketLike;
  signal?: AbortSignal;
}

const DEFAULT_URL = "wss://api.openai.com/v1/realtime";
const OPEN = 1;

/**
 * WebSocket transport, where audio is base64 PCM inside JSON events.
 *
 * WebRTC is the better choice on a phone or in a browser because it handles packet loss and jitter,
 * but WebSocket needs no media stack at all, which makes it the transport that works everywhere:
 * servers, tests, terminals, and any platform without a WebRTC dependency.
 */
export async function connectWebSocket(options: WebSocketTransportOptions): Promise<RealtimeTransport> {
  const credentials = await options.credentials.get();
  const url = endpointWithModel(options.url ?? DEFAULT_URL, options.model);

  const protocols = [
    "realtime",
    `openai-insecure-api-key.${credentials.token}`,
    ...(options.organization ? [`openai-organization.${options.organization}`] : []),
    ...(options.project ? [`openai-project.${options.project}`] : []),
  ];

  const factory =
    options.createWebSocket ??
    ((target: string, subprotocols: string[]) => {
      const Ctor = (globalThis as { WebSocket?: new (url: string, protocols?: string[]) => WebSocketLike }).WebSocket;
      if (!Ctor) throw new Error("no WebSocket implementation available; pass createWebSocket");
      return new Ctor(target, subprotocols);
    });

  const socket = factory(url, protocols);

  const messageListeners = new Set<(raw: Record<string, unknown>) => void>();
  const closeListeners = new Set<(reason?: string) => void>();
  const errorListeners = new Set<(error: Error) => void>();

  socket.addEventListener("message", (event: { data: unknown }) => {
    const text = typeof event.data === "string" ? event.data : String(event.data);
    let parsed: unknown;
    try {
      parsed = JSON.parse(text);
    } catch {
      for (const listener of errorListeners) listener(new Error("received a message that was not JSON"));
      return;
    }
    if (typeof parsed === "object" && parsed !== null) {
      for (const listener of messageListeners) listener(parsed as Record<string, unknown>);
    }
  });

  socket.addEventListener("close", (event: { reason?: string; code?: number }) => {
    for (const listener of closeListeners) listener(event?.reason || (event?.code ? `code ${event.code}` : undefined));
  });

  socket.addEventListener("error", () => {
    for (const listener of errorListeners) listener(new Error("the realtime socket failed"));
  });

  await new Promise<void>((resolve, reject) => {
    if (socket.readyState === OPEN) return resolve();
    const onAbort = () => reject(new Error("connection aborted"));
    options.signal?.addEventListener("abort", onAbort, { once: true });
    socket.addEventListener("open", () => {
      options.signal?.removeEventListener("abort", onAbort);
      resolve();
    });
    socket.addEventListener("error", () => reject(new Error("could not open the realtime socket")));
    socket.addEventListener("close", (event: { reason?: string }) =>
      reject(new Error(`the realtime socket closed before it opened${event?.reason ? `: ${event.reason}` : ""}`)),
    );
  });

  return {
    kind: "websocket",
    carriesAudioOutOfBand: false,

    send(message) {
      if (socket.readyState !== OPEN) throw new Error("the realtime socket is not open");
      socket.send(JSON.stringify(message));
    },

    sendAudio(pcm) {
      if (socket.readyState !== OPEN) return;
      socket.send(JSON.stringify({ type: CLIENT_EVENTS.appendAudio, audio: encodeBase64(pcm) }));
    },

    onMessage(listener) {
      messageListeners.add(listener);
    },
    onClose(listener) {
      closeListeners.add(listener);
    },
    onError(listener) {
      errorListeners.add(listener);
    },

    async close(reason) {
      if (socket.readyState === OPEN) socket.close(1000, reason?.slice(0, 120));
    },
  };
}
